src/Features/Core/Portable/Diagnostics/BuildOnlyDiagnosticsService.cs

```diff
@@ -0,0 +1,147 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    [ExportWorkspaceServiceFactory(typeof(IBuildOnlyDiagnosticsService), ServiceLayer.Default), Shared]
+    internal sealed class BuildOnlyDiagnosticsServiceFactory : IWorkspaceServiceFactory
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public BuildOnlyDiagnosticsServiceFactory()
+        {
+        }
+
+        public IWorkspaceService CreateService(HostWorkspaceServices workspaceServices)
+            => new BuildOnlyDiagnosticsService(workspaceServices.Workspace);
+
+        private sealed class BuildOnlyDiagnosticsService : IBuildOnlyDiagnosticsService
+        {
+            private readonly object _gate = new();
+            private readonly Dictionary<DocumentId, ImmutableArray<DiagnosticData>> _documentDiagnostics = new();
+            private readonly Dictionary<ProjectId, ImmutableArray<DiagnosticData>> _projectDiagnostics = new();
+
+            public BuildOnlyDiagnosticsService(Workspace workspace)
+            {
+                workspace.WorkspaceChanged += OnWorkspaceChanged;
+            }
+
+            private void OnWorkspaceChanged(object? sender, WorkspaceChangeEventArgs e)
```

Comment:
My expectation is that once the workspace is changed in any way then the diagnostics are stale. In practice I would expect the set of build-only diagnostics to be less likly to be impacted by a document change but it is a question of how long we want stale diagnostics to stick around for.

There seems to be two possibilities here:

1. You build, get a diagnostics and they don't do away until you either rebuild or unload/reload the document that they were trigger for (or project if they are project level diagnostics)
2. You build, get a diagnostics and they go away as soon as anything changes.

(2) seems bad because then the diagnostics are too ephemeral. (1) seems fine since I want the diagnostics to stick around at least long enough for me to fix them. Ideally we could track the location of build-only diagnostics and remove them when edits to a file overlap them. My main concern is that its not clear to users that they will need to do another build to clear out build-only diagnostics.

Reply:
Thanks, I filed https://github.com/dotnet/roslyn/issues/64659 to track this work. We have seen bunch of reports already about stale build-only diagnostics after the code has been updated or fixed. For example, unused field compiler diagnostic hangs around after you remove the field or add a read/write to the field. We definitely need to improve the experience here.